ocaml/xapi/xapi_vbd.ml

```diff
@@ -47,10 +77,23 @@ let plug ~__context ~self =
     Storage_access.attach_and_activate ~__context ~vbd:self ~domid
       (fun attach_info ->
          let params = attach_info.Storage_interface.params in
-         let prefix = "/dev/" in
-         let prefix_len = String.length prefix in
-         let path = String.sub params prefix_len (String.length params - prefix_len) in
-         Db.VBD.set_device ~__context ~self ~value:path;
+         let device_path =
+           let nbd_prefix = "hack|nbd:unix:" in
+           let is_nbd = String.startswith nbd_prefix params in
+           if is_nbd then begin
+             let export_name = "qemu_node" in
+             let unix_socket_path = params |> Astring.String.cuts ~empty:false ~sep:nbd_prefix |> List.hd |> String.split_on_char '|' |> List.hd in
```

Comment:
This will raise an horrible looking and hard to debug error if there are empty lists on the way

Reply:
Yes, but it's isolated

Reply:
Maybe I could add a debug message above this with the content of `params`, so that we now what we got in case one of the `List.hd`s fail?

Reply:
I think adding a preliminary debug with the params could be enough.
The only thing with List.hd is probably to define your own `hd_log` as something like
```ocaml
let hd_log logline = function
| h :: _ -> h
| [] -> debug logline; raise Not_found (* or whatever else List.hd raises *)
```

Reply:
Thanks, I've added a debug line.